### PR TITLE
Validate views when publishing

### PR DESCRIPTION
**Pull request by tpluscode**
This PR adds a bunch of changes which greatly improve view publishing:

1. Validation, which for now checks the view metadata (fetched from the live endpoint)
2. In case of errors, the UI can display a summary of problems
3. Validation checks can be marked as warnings. Warnings can be ignored when publishing

[https_view-builder.lndo.siteappviews.webm](https://user-images.githubusercontent.com/588128/199235659-e5813a4e-89a7-469a-b9c9-464e028a8cc6.webm)

### CODE REVIEW

File: apps/www/src/element/ssz-shacl-report.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import { html, LitElement } from 'lit'
+import TermMap from '@rdfjs/term-map'
+import { sh } from '@tpluscode/rdf-ns-builders'
+import '@shoelace-style/shoelace/dist/components/details/details.js'
+import { toSparql } from 'clownface-shacl-path'
+
+customElements.define('ssz-shacl-report', class extends LitElement {
```
Comment (tpluscode) [inline apps/www/src/element/ssz-shacl-report.js:7]: This is the dialog element which displays shacl reports

File: apps/www/src/state/plugin/notifications.js [hunk 2]
```diff
@@ -23,6 +23,15 @@ export const notifications = {
...
         return state
       },
+      errorDialog(state, error) {
+        return { ...state, error }
+      },
+      hideError(state) {
+        return {
+          ...state,
+          error: undefined,
+        }
```
Comment (tpluscode) [inline apps/www/src/state/plugin/notifications.js:33]: The SHACL dialog is displayed when the `error` state is set and hidden when undefined

File: apps/www/src/state/plugin/operationAlerts.js [hunk 3]
```diff
@@ -1,15 +1,35 @@
...
+          if (error) {
+            const { pointer } = error
+            dispatch.notifications.show({
+              content: html`
+                <strong>${pointer.out(rdfs.label).value}</strong><br>
+                <sl-button outline @click="${() => dispatch.notifications.errorDialog(pointer)}">See details</sl-button>
+              `,
+              variant: 'danger',
+              autoHide: false,
+            })
```
Comment (tpluscode) [inline apps/www/src/state/plugin/operationAlerts.js:26]: See how a notification can have rich content, with action trigger for opening the SHACL details

File: packages/publish-views/lib/ValidationError.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import { rdf, sh } from '@tpluscode/rdf-ns-builders'
+import $rdf from 'rdf-ext'
+import clownface from 'clownface'
+
+export class ValidationError extends Error {
```
Comment (tpluscode) [inline packages/publish-views/lib/ValidationError.js:5]: This is similar to the error class from barnard package, but I needed something different, which would combine multiple reports

File: packages/publish-views/lib/loadViews.js [hunk 5]
```diff
@@ -4,26 +4,46 @@ import through2 from 'through2'
...
 }
 
+async function loadViewMeta(view, client) {
+  const shape = await viewShape()
+
+  const subjectVariable = 'view'
+  const query = shapeTo.construct(shape, { focusNode: view, subjectVariable })
+
+  return query.execute(client.query)
+}
```
Comment (tpluscode) [inline packages/publish-views/lib/loadViews.js:45]: Here, the metadata endpoint is queries for the properties of a view so that it can be validated in int entirety

File: packages/publish-views/lib/shacl.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+export function combineShaclReports({ context, report }) {
+  const variables = context.variables.has(REPORTS_KEY)
+    ? context.variables
+    : context.variables.set(REPORTS_KEY, [])
+
+  variables.get(REPORTS_KEY).push(report)
+
+  // continue stream
+  return true
+}
```
Comment (tpluscode) [inline packages/publish-views/lib/shacl.js:16]: A callback to `barnard59-validate-shacl` to continue processing all of the views.

It will only kill the pipeline when everything is done (see below)

File: packages/publish-views/lib/shacl.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    this.push(chunk)
+    next()
+  }, function (done) {
+    const reports = variables.get(REPORTS_KEY)
+    if (reports?.some(hasViolations(ignoreWarnings))) {
+      this.destroy(new ValidationError(reports))
+    }
+
+    done()
+  })
```
Comment (tpluscode) [inline packages/publish-views/lib/shacl.js:32]: This step is basically a no-op. The flush callback destroys the stream if there are error in the end and also has the option to ignore the reports if they contain only `sh:Warning`

File: packages/publish-views/shapes.ttl [hunk 8]
```diff
@@ -0,0 +1,96 @@
+PREFIX dcterms: <http://purl.org/dc/terms/>
+PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
+PREFIX schema: <http://schema.org/>
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX sh: <http://www.w3.org/ns/shacl#>
+prefix view: <https://cube.link/view/>
+
+<ViewShape>
```
[the remark this answers is not in the document]
Comment (tpluscode) [reply packages/publish-views/shapes.ttl:10]: This shape is used to validate view meta and also to generate a `CONSTRUCT` to fetch view metadata

File: apps/api/lib/errors.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+      detail: error.message,
+      status: 400,
+    }, {
+      '@context': {
+        '@vocab': hex().value,
+      },
+      report,
+    })
+
+    problem.title = 'Data issues were found'
```
Comment (cristianvasquez) [inline apps/api/lib/errors.js:25]: I seems 'Data issues were found' is already in the problem instance

File: apps/api/lib/viewCollection/publish.js [hunk 10]
```diff
@@ -4,39 +4,62 @@ import { viewBuilder } from '@view-builder/core/ns.js'
...
   }
 })
 
-async function downloadViews(req, res) {
-  const { stream } = await publishViews.toNtriples(req.labyrinth.sparql)
-
-  res.setHeader('Content-type', 'text/plain')
-  res.setHeader('Content-Disposition', 'attachment; filename=views.nt')
+function downloadViews(req, res, next, ignoreWarnings) {
+  temporaryFileTask(async (temporaryPath) => {
```
[the remark this answers is not in the document]
Comment (cristianvasquez) [reply apps/api/lib/viewCollection/publish.js:23]: Maybe a worker pool with limited parallel processes? 

Using a lib: https://www.npmjs.com/package/worker-farm (I've just quickly found that one, I'm not aware of the good libraries that implement this in node)

I think it's an option if the performance decays when later. Not right now.

File: apps/www/src/element/ssz-shacl-report.js (hunk 1 above)
Comment (cristianvasquez) [reply apps/www/src/element/ssz-shacl-report.js:7]: This is probably a good candidate to be in a library somewhere. It's quite common to show validation reports.

I'm not sure which lib though

File: packages/publish-views/shapes.ttl [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+      sh:severity sh:Warning ;
+    ],
+    [
+      sh:path schema:description ;
+      sh:nodeKind sh:Literal ;
+      sh:severity sh:Warning ;
+    ],
+    [
+      sh:path schema:endDate ;
+      sh:datatype xsd:date ;
```
Comment (cristianvasquez) [inline packages/publish-views/shapes.ttl:56]: sh:maxCount 1; ?

File: packages/publish-views/shapes.ttl [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+      sh:or
+        (
+          [ sh:datatype xsd:date ]
+          [ sh:datatype xsd:dateTime ]
+        ) ;
+      sh:severity sh:Warning ;
+    ],
+    [
+      sh:path dcterms:publisher ;
+      sh:nodeKind sh:IRI ;
```
Comment (cristianvasquez) [inline packages/publish-views/shapes.ttl:41]: min 0, max 1

File: packages/publish-views/shapes.ttl [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+      sh:severity sh:Warning ;
+    ],
+    [
+      sh:path schema:alternateName ;
+      sh:nodeKind sh:Literal ;
+      sh:severity sh:Warning ;
+    ],
+    [
+      sh:path schema:description ;
+      sh:nodeKind sh:Literal ;
```
Comment (cristianvasquez) [inline packages/publish-views/shapes.ttl:51]: min 0, max 1 . 
If it's the case

File: apps/www/src/element/ssz-shacl-report.js (hunk 1 above)
Comment (tpluscode) [reply apps/www/src/element/ssz-shacl-report.js:7]: There is no such lib right now. Could put it in the https://github.com/zazuko/rdfjs-elements/ repo (even if a new package)

Will create an issue

File: packages/publish-views/shapes.ttl (hunk 11 above)
Comment (tpluscode) [reply packages/publish-views/shapes.ttl:56]: I wanted to keep this shape fairly permissive for now but I `maxCount` sounds good

File: packages/publish-views/shapes.ttl (hunk 13 above)
Comment (tpluscode) [reply packages/publish-views/shapes.ttl:51]: In fact, `maxCount` may not be a good fit here in case we want multilingual annotations. I don't know, hence I kept this mostly unspecified

File: packages/publish-views/shapes.ttl (hunk 12 above)
Comment (tpluscode) [reply packages/publish-views/shapes.ttl:41]: Don't know, actually :)

File: packages/publish-views/lib/shacl.js (hunk 6 above)
Comment (cristianvasquez) [reply packages/publish-views/lib/shacl.js:16]: Interesting functionality, I wonder if this should go as a functionality of `barnard59-validate-shacl` itself, what do you think?

File: packages/publish-views/lib/shacl.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
+import through2 from 'through2'
+import { sh } from '@tpluscode/rdf-ns-builders'
+import { ValidationError } from './ValidationError.js'
+
+const REPORTS_KEY = 'COMBINED_SHACL_REPORTS'
+
+export function combineShaclReports({ context, report }) {
```
Comment (cristianvasquez) [inline packages/publish-views/lib/shacl.js:7]: I would call it 'collectShaclReports'

File: packages/publish-views/lib/shacl.js (hunk 6 above)
Comment (tpluscode) [reply packages/publish-views/lib/shacl.js:16]: Maybe. I did something a little similar in cube creator too because I also did not want the pipeline to be stopped on first error

File: packages/publish-views/lib/loadViews.js [hunk 15]
```diff
@@ -4,26 +4,46 @@ import through2 from 'through2'
...
-      ?view a ${ns.view.View} ; ${ns.viewBuilder.publish} true ;
+      ?view 
+        a ${ns.view.View} ;
+        ${schema.sameAs} ?destination ; 
+        ${ns.viewBuilder.publish} true ;
     `
     .execute(client.query)
 
-  return views.pipe(through2.obj(async function ({ view }, _, next) {
+  return views.pipe(through2.obj(async function ({ view, destination }, _, next) {
```
Comment (cristianvasquez) [inline packages/publish-views/lib/loadViews.js:26]: The word 'destination' doesn't tell me much, is it something like metadataIri ?

File: packages/publish-views/lib/shacl.js (hunk 6 above)
Comment (cristianvasquez) [reply packages/publish-views/lib/shacl.js:16]: I think it's a very reasonable functionality.  

I'm unsure about one REPORTS_KEY if there are many validator instances in the pipeline though

**Review by cristianvasquez — CHANGES_REQUESTED** [review]
It looks good to me, I would only add a snapshot-based test to cover the edge cases.

File: packages/publish-views/lib/loadViews.js (hunk 15 above)
Comment (cristianvasquez) [reply packages/publish-views/lib/loadViews.js:26]: perhaps something along the lines of draftUri and publishedUri ?

File: apps/www/src/element/ssz-shacl-report.js (hunk 1 above)
Comment (cristianvasquez) [reply apps/www/src/element/ssz-shacl-report.js:7]: That would be awesome; I've also seen other apps that don't know how to show such a report. It's nice to have it centralized

**Comment by tpluscode** [comment]
I applied the small changes. Tests in progress in #45
